test/e2e/restart.go

```diff
@@ -0,0 +1,367 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// How long each node is given during a process that restarts all nodes
+	// before the test is considered failed. (Note that the total time to
+	// restart all nodes will be this number times the nubmer of nodes.)
+	restartPerNodeTimeout = 5 * time.Minute
```

Comment:
Check out this line and the comment I wrote above it:
https://github.com/GoogleCloudPlatform/kubernetes/pull/8837/files#diff-bbc40d90a96515627afa5efea9599dbeR292

We can just set this to the cluster size if we want it to run in parallel. Then, the whole thing should take roughly a single node's time.

I originally wrote it this way because during node upgrades, we expect to do it serially, so that if something fails, we destroy the minimum number of nodes we can :-)

However, for testing, we could either have a flag to do it serially for "normal" e2es and in parallel for scalability runs, or just always do it in parallel as it's probably going to be roughly the same test.

Another thought: for "soak" tests, should we be disabling all reboot / restart / recreate tests? So that we see how long the system really lasts?